### PR TITLE
Adding merge function as an optional input for KeyZipper

### LINKED ISSUE #44: KeyZipper improvement
Currently multiple stacked `KeyZipper` would create a recursive data structure:
```py
dp = KeyZipper(dp, ref_dp1, lambda x: x)
dp = KeyZipper(dp, ref_dp2, lambda x: x[0])
dp = KeyZipper(dp, ref_dp3, lambda x: x[0][0])
```
This is super annoying if we are using same key for each `KeyZipper`. At the end, it yields `(((dp, ref_dp1), ref_dp2), ref_dp3)

We should either accept multiple reference DataPipe for KeyZipper to preserve same key, or have some expand or collate function to convert result to `(dp, (ref_dp1, ref_dp2, ref_dp3))`

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76

Fixes #44 by adding merge function as an optional input for `KeyZipper`. It allows user to specify how the items yielded from the DataPipes will be combined before they are yielded.

This PR changes the default behavior when `keep_key=True` and that may be BC breaking.
Previously, when `keep_key=True`, the output will be `(key, item1, item2)`, now it will default to `(key, (item1, item2)`. It is possible to leave the default behavior unchanged.

Note that `KeyZipper` may be renamed to `IterZipper` in #50 to provide users with better clarity.

Differential Revision: [D31820203](https://our.internmc.facebook.com/intern/diff/D31820203)

### CODE REVIEW

File: test/test_datapipe.py [hunk 1]
```diff
@@ -103,7 +103,27 @@ def test_keyzipper_iterdatapipe(self) -> None:
         zip_dp_w_key = source_dp.zip_by_key(
             ref_datapipe=ref_dp, key_fn=lambda x: x, ref_key_fn=lambda x: x, keep_key=True, buffer_size=10
         )
-        self.assertEqual([(i, i, i) for i in range(10)], list(zip_dp_w_key))
+        self.assertEqual([(i, (i, i)) for i in range(10)], list(zip_dp_w_key))
```
Comment (NivekT) [inline test/test_datapipe.py:106]: Note the change in default behavior here.

File: test/test_datapipe.py [hunk 2]
```diff
@@ -103,7 +103,27 @@ def test_keyzipper_iterdatapipe(self) -> None:
...
+        # Functional Test: using a different merge function and keep_key=True
+        zip_dp_w_key_str_merge = source_dp.zip_by_key(
+            ref_datapipe=ref_dp,
+            key_fn=lambda x: x,
+            ref_key_fn=lambda x: x,
+            keep_key=True,
+            buffer_size=10,
+            merge_fn=merge_to_string,
+        )
+        self.assertEqual([(i, f"{i},{i}") for i in range(10)], list(zip_dp_w_key_str_merge))
```
Comment (ejguan) [inline test/test_datapipe.py:127]: Can we add a test as an example to show unwrap the nested list?

File: torchdata/datapipes/iter/util/combining.py [hunk 3]
```diff
@@ -6,31 +6,46 @@
...
     def __init__(
-        self, source_datapipe, ref_datapipe, key_fn, ref_key_fn=None, keep_key=False, buffer_size=10000,
+        self,
+        source_datapipe: IterDataPipe,
+        ref_datapipe: IterDataPipe,
+        key_fn: Callable,
+        ref_key_fn: Callable = None,
+        keep_key: bool = False,
+        buffer_size: int = 10000,
+        merge_fn: Callable = tuple_merge,
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/combining.py:41]: nit: Can we make it None as default?

File: torchdata/datapipes/iter/util/combining.py [hunk 4]
```diff
@@ -6,31 +6,46 @@
...
     ):
         self.source_datapipe = source_datapipe
         self.ref_datapipe = ref_datapipe
         self.key_fn = key_fn
         self.ref_key_fn = key_fn if ref_key_fn is None else ref_key_fn
         self.keep_key = keep_key
+        self.merge_fn = merge_fn
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/combining.py:48]:
```suggestion
        self.merge_fn = tuple if merge_fn is None else merge_fn
```
WDYT?

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31820203).

**Pull request closed without merging**